src/node/dynamics_compressor.rs

```diff
@@ -0,0 +1,512 @@
+            let ring_buffer_size =
+                (context.sample_rate() * 0.006 / RENDER_QUANTUM_SIZE as f32).ceil() as usize + 1;
+            let ring_buffer = Vec::<AudioRenderQuantum>::with_capacity(ring_buffer_size);
+
+            let render = DynamicsCompressorRenderer {
+                attack: attack_proc,
+                knee: knee_proc,
+                ratio: ratio_proc,
+                release: release_proc,
+                threshold: threshold_proc,
+                reduction: reduction.clone(),
+                ring_buffer,
+                ring_index: 0,
+                prev_detector_value: 0.,
+            };
+
+            let node = DynamicsCompressorNode {
+                registration,
+                channel_config: options.channel_config.into(),
+                attack: attack_param,
+                knee: knee_param,
+                ratio: ratio_param,
+                release: release_param,
+                threshold: threshold_param,
+                reduction,
+            };
+
+            (node, Box::new(render))
+        })
+    }
+
+    pub fn attack(&self) -> &AudioParam {
+        &self.attack
+    }
+
+    pub fn knee(&self) -> &AudioParam {
+        &self.knee
+    }
+
+    pub fn ratio(&self) -> &AudioParam {
+        &self.ratio
+    }
+
+    pub fn release(&self) -> &AudioParam {
+        &self.release
+    }
+
+    pub fn threshold(&self) -> &AudioParam {
+        &self.threshold
+    }
+
+    pub fn reduction(&self) -> f32 {
+        self.reduction.load()
+    }
+}
+
+struct DynamicsCompressorRenderer {
+    attack: AudioParamId,
+    knee: AudioParamId,
+    ratio: AudioParamId,
+    release: AudioParamId,
+    threshold: AudioParamId,
+    reduction: Arc<AtomicF32>,
+    ring_buffer: Vec<AudioRenderQuantum>,
+    ring_index: usize,
+    prev_detector_value: f32,
+}
+
+// SAFETY:
+// AudioRenderQuantums are not Send but we promise the `ring_buffer` Vec is
+// empty before we ship it to the render thread.
+#[allow(clippy::non_send_fields_in_send_ty)]
+unsafe impl Send for DynamicsCompressorRenderer {}
+
+// https://webaudio.github.io/web-audio-api/#DynamicsCompressorOptions-processing
+// see also https://www.eecs.qmul.ac.uk/~josh/documents/2012/GiannoulisMassbergReiss-dynamicrangecompression-JAES2012.pdf
+// follow Fig. 7 (c) diagram in paper
+impl AudioProcessor for DynamicsCompressorRenderer {
+    fn process(
+        &mut self,
+        inputs: &[AudioRenderQuantum],
+        outputs: &mut [AudioRenderQuantum],
+        params: AudioParamValues,
+        scope: &RenderScope,
+    ) -> bool {
+        // single input/output node
+        let input = inputs[0].clone();
+        let output = &mut outputs[0];
+        let sample_rate = scope.sample_rate;
+
+        let ring_size = self.ring_buffer.capacity();
+        // ensure ring buffer is filled with silence
+        if self.ring_buffer.len() < ring_size {
+            let mut silence = input.clone();
```

Comment:
You can call `input.silence()` to create a new silent buffer

Reply:
Actually no, the `silence()` method is only on the `AudioRenderQuantumChannel`

But I agree having to write these two lines is a bit cumbersome. Maybe `make_silent()` could return `self` so that we could write in one line:

```rs
let silence = channel.clone().make_silent();
```

Which I think looks like a good tradeoff?